tools/fixedpointarithmetic/src/main/java/dk/alexandra/fresco/decimal/DefaultAdvancedRealNumeric.java

```diff
@@ -0,0 +1,194 @@
+      RealNumeric fixed = provider.apply(par);
+      List<DRes<SReal>> products = new ArrayList<>(a.size());
+      for (int i = 0; i < a.size(); i++) {
+        products.add(fixed.numeric().mult(a.get(i), b.get(i)));
+      }
+
+      return () -> products;
+    }).seq((seq, list) -> {
+      RealNumeric fixed = provider.apply(seq);
+      return fixed.advanced().sum(list);
+    });
+  }
+
+  @Override
+  public DRes<SReal> innerProductWithPublicPart(List<BigDecimal> a, List<DRes<SReal>> b) {
+    return builder.par(r1 -> {
+      if (a.size() != b.size()) {
+        throw new IllegalArgumentException("Vectors must have same size");
+      }
+
+      RealNumeric fixed = provider.apply(r1);
+      List<DRes<SReal>> products = new ArrayList<>(a.size());
+      for (int i = 0; i < a.size(); i++) {
+        products.add(fixed.numeric().mult(a.get(i), b.get(i)));
+      }
+
+      return () -> products;
+    }).seq((seq, list) -> {
+      RealNumeric fixed = provider.apply(seq);
+      return fixed.advanced().sum(list);
+    });
+  }
+
+  @Override
+  public DRes<SReal> exp(DRes<SReal> x) {
+
+    int numberOfTerms = 16;
+    return builder.seq(seq -> {
+      List<DRes<SReal>> powers = new ArrayList<>(numberOfTerms - 1);
+
+      powers.add(x);
+      DRes<SReal> pow = x;
+      RealNumeric fixed = provider.apply(seq);
+      for (int i = 1; i < numberOfTerms - 1; i++) {
+        pow = fixed.numeric().mult(pow, x);
+        powers.add(pow);
+      }
+      return () -> powers;
+    }).par((par, powers) -> {
+
+      /*
+       * We approximate the exponential function by calculating the first terms of the Taylor
+       * expansion. By letting all terms in the series have common denominator, we only need to do
+       * one division.
+       * 
+       * TODO: In the current implementation we compute 16 terms, which seems to give decent
+       * percision for small inputs. If we want full precision for all possible inputs, we need to
+       * calculate many more terms, since the error after n terms is approx x^{n+1}/(n+1)!), and
+       * this would cause the function to be very inefficient. Maybe we should allow the app
+       * developer to specify an upper bound on the input?
+       */
+
+      BigInteger n = BigInteger.ONE;
+      List<DRes<SReal>> terms = new ArrayList<>(numberOfTerms);
+      RealNumeric fixed = provider.apply(par);
+      for (int i = numberOfTerms - 1; i >= 1; i--) {
+        terms.add(fixed.numeric().mult(new BigDecimal(n), powers.get(i - 1)));
+        n = n.multiply(BigInteger.valueOf(i));
+      }
+      final BigDecimal divisor = new BigDecimal(n);
+      terms.add(fixed.numeric().known(divisor));
+      return () -> new Pair<>(terms, divisor);
+    }).seq((seq, termsAndDivisor) -> {
+      RealNumeric fixed = provider.apply(seq);
+      return fixed.numeric().div(fixed.advanced().sum(termsAndDivisor.getFirst()),
+          termsAndDivisor.getSecond());
+    });
+  }
+
+  @Override
+  public DRes<SReal> random() {
```

Comment:
This looks like a very dangerous way to implement the random method: 
1. The `DEFAULT_RANDOM_BITS` of 32 bits is not very much in a cryptographic setting  
2. This is not documented in the Javadoc of the API, so there is not way for a user to know that this is the case. The javadoc just says a "random number between 0 and 1". This should be made much more precise so the user knows how much entropy to except.
3. Implementing it using a random additive mask is very costly as each bit requires at least a single triple in the preprocessing phase.

While 3. may be unavoidable, 1. and 2. really bad and could very well damage the security of computations build using this method.